x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go

```diff
@@ -0,0 +1,215 @@
+			ModuleFields:    moduleFields,
+			MetricSetFields: metricsetFields,
+		}); !isReported {
+			m.log.Debug("event not reported")
+		}
+
+	}
+}
+
+// Close the connection to the server at the engine level
+func (m *MetricSet) Close() error {
+	return m.db.Close()
+}
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (common.MapStr, error) {
+	// According to MS docs a single result is always returned for this query
+	row := m.db.QueryRow(fmt.Sprintf(`USE %s; SELECT * FROM sys.dm_db_log_space_usage;`, dbName))
+
+	var res logSpace
+	if err := row.Scan(&res.id, &res.totalLogSizeInBytes, &res.usedLogSpaceInBytes, &res.usedLogSpaceInPercent,
+		&res.logSpaceInBytesSinceLastBackup); err != nil {
+		// Because this query only returns a single result an error in the first scan is
+		// probably a "data returned but not properly scanned"
+		err = errors.Wrap(err, "error scanning single result")
+		return nil, err
+	}
+
+	return common.MapStr{
+		"total": common.MapStr{
+			"bytes": res.totalLogSizeInBytes,
+		},
+		"used": common.MapStr{
+			"bytes": res.usedLogSpaceInBytes,
+			"pct":   res.usedLogSpaceInPercent,
+		},
+		"since_last_backup": common.MapStr{
+			"bytes": res.logSpaceInBytesSinceLastBackup,
+		},
+	}, nil
+}
+func (m *MetricSet) getLogStats(db dbInfo) (common.MapStr, error) {
+	// According to MS docs a single result is always returned for this query
+	row := m.db.QueryRow(fmt.Sprintf(`USE %s; SELECT database_id,total_log_size_mb,active_log_size_mb,log_backup_time,log_since_last_log_backup_mb,log_since_last_checkpoint_mb,log_recovery_size_mb FROM sys.dm_db_log_stats(%d);`, db.name, db.id))
+
+	var res logStats
+	if err := row.Scan(&res.databaseID, &res.sizeMB, &res.activeSizeMB, &res.backupTime, &res.sinceLastBackupMB, &res.sinceLastCheckpointMB, &res.recoverySizeMB); err != nil {
+		// Because this query only returns a single result an error in the first scan is
+		// probably a "data returned but not properly scanned"
+		err = errors.Wrap(err, "error scanning single result")
+		return nil, err
+	}
+
+	result := common.MapStr{
+		"total_size": common.MapStr{
+			"bytes": res.sizeMB * 1048576,
+		},
+		"active_size": common.MapStr{
+			"bytes": res.activeSizeMB * 1048576,
+		},
+		"backup_time": res.backupTime,
+		"since_last_checkpoint": common.MapStr{
+			"bytes": res.sinceLastCheckpointMB * 1048576,
+		},
+		"recovery_size": common.MapStr{
+			"bytes": res.recoverySizeMB,
+		},
+	}
+
+	if res.sinceLastBackupMB != nil {
+		result.Put("since_last_backup.bytes", *res.sinceLastBackupMB*1048576)
+	}
+
+	return result, nil
+}
+
+func (m *MetricSet) getDbsNames() (res []dbInfo, err error) {
+	res = make([]dbInfo, 0)
+
+	var rows *sql.Rows
+	rows, err = m.db.Query("SELECT name, database_id FROM sys.databases")
+	if err != nil {
+		return
+	}
+	defer closeRows(rows)
+
+	for rows.Next() {
+		var row dbInfo
+		if err = rows.Scan(&row.name, &row.id); err != nil {
+			return
+		}
+
+		res = append(res, row)
+	}
+
+	return
```

Comment:
named returns?

Reply:
we can also delay this, no needed in this PR.

Reply:
All of them!?!? Oookkk :roll_eyes: Done